source/common/access_log/access_log_formatter.h

```diff
@@ -126,5 +130,35 @@ class RequestInfoFormatter : public Formatter {
   std::function<std::string(const RequestInfo::RequestInfo&)> field_extractor_;
 };
 
+/**
+ * Base formatter for formatting Metadata objects
+ */
+class MetadataFormatter {
+public:
+  MetadataFormatter(const std::string& filter_namespace, const std::vector<std::string>& path,
+                    const absl::optional<size_t>& max_length);
+
+  std::string format(const envoy::api::v2::core::Metadata& metadata) const;
+
+private:
+  std::string filter_namespace_;
+  std::vector<std::string> path_;
+  absl::optional<size_t> max_length_;
+};
+
+/**
+ * Formatter based on the DynamicMetadata from RequestInfo.
+ */
+class DynamicMetadataFormatter : public Formatter, MetadataFormatter {
+public:
+  DynamicMetadataFormatter(const std::string& filter_namespace,
+                           const std::vector<std::string>& path,
+                           const absl::optional<size_t>& max_length);
```

Comment:
It's probably not worth passing this one by ref, it will be a tiny object that can fit in a register.

Reply:
It's the same pattern as used in the other formatters: ResponseHeaderFormatter/RequestHeaderFormatter.